### PR TITLE
 Fixed: App can only be accessed on localhost

### LINKED ISSUE #1244: App can only be accessed on `localhost`
**Describe the bug**
1. The codebase hard codes localhost in many places
1. This means you cannot get Talawa-Admin to work while being accessed from a remote host without a proxy

**To Reproduce**

1. Install  Talawa-Admin on a remote system
1. Errors are seen when accessing the app directly from a remote host

**Expected behavior**
1. Access must be available from a remote host that is not located on `localhost`. This can only be truly tested when run from a completely different system.
1. The installation guide must show how to access Admin from a remote system
1. All hard coded instances of `localhost` must be replaced by configuration values in `.env`  and/or the browser URL where applicable

**Actual behavior**
- See screenshots

**Screenshots**

1. The `.env` file
image
1. Devtools output
image
1. The codebase hard codes localhost in many places
image

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by aarishshahmohsin**
**What kind of change does this PR introduce?**

This PR fixes the problem of the portal not being able to be accessed if not on localhost.

**Issue Number:**

Fixes #1244 

**Did you add tests for your changes?**

Yes

**Snapshots/Videos:**

IMG_1534

**If relevant, did you update the documentation?**

Yes

**Summary**

This issue was caused mainly by the case in which the API url contains 'localhost'.

**Does this PR introduce a breaking change?**

No

**Other information**
2 problems occurred while working on the issue.
1. One problem that needs to be fixed is the plugin store, which is coded incorrectly. in the src/components/AddOn/support/services/Plugin.helper.ts file.

```
    const result = await fetch(`http://localhost:${process.env.PORT}/store`);
    const result = await fetch(`http://localhost:3005/installed`);
```

are both fetching endpoints that do not exist and hence the addon store component also does not work. 

2. The navbar button is not working on mobile devices.

Should I make issues for them?

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

Yes

### CODE REVIEW

File: setup.ts [hunk 1]
```diff
@@ -48,6 +49,39 @@ export async function main(): Promise<void> {
...
+    );
+    shouldSetRemoteHost = true;
+  } else {
+    const { shouldSetRemoteHostResponse } = await inquirer.prompt({
+      type: 'confirm',
+      name: 'shouldSetRemoteHostResponse',
+      message: 'Would you like to set up a remote host?',
+      default: true,
+    });
+    shouldSetRemoteHost = shouldSetRemoteHostResponse;
```
Comment (palisadoes) [inline setup.ts:66]: 1. This is unnecessary, setup already asks for the `REACT_APP_TALAWA_URL`  value that includes a TCP port. BTW, You cannot assume the port will never change. 
2. You are duplicating the prompting for an existing parameter. You should be able to derive the necessary information from the `REACT_APP_TALAWA_URL`  value.

File: src/components/Advertisements/core/AdvertisementEntry/AdvertisementEntry.test.tsx [hunk 2]
```diff
@@ -308,43 +308,6 @@ describe('Testing Advertisement Entry Component', () => {
...
   });
 
-  test('Updates the advertisement and shows error toast on successful update', async () => {
-    const updateAdByIdMock = jest.fn();
-
-    mockUseMutation.mockReturnValue([updateAdByIdMock]);
-
-    render(
-      <ApolloProvider client={client}>
-        <Provider store={store}>
```
Comment (palisadoes) [inline src/components/Advertisements/core/AdvertisementEntry/AdvertisementEntry.test.tsx:318]: Why was this removed?

File: src/setup/askForRemoteUrl/askForRemoteUrl.test.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  test('should return the provided endpoint when user enters it', async () => {
+    const mockPrompt = jest.spyOn(inquirer, 'prompt').mockResolvedValueOnce({
+      url: 'http://example.com/',
+    });
+    const result = await askForRemoteUrl();
+    expect(mockPrompt).toHaveBeenCalledWith([
+      {
+        type: 'input',
+        name: 'url',
+        message: `Enter your remote url in the format 'http://hosturl/' (don't enter the port)`,
```
Comment (palisadoes) [inline src/setup/askForRemoteUrl/askForRemoteUrl.test.ts:32]: This not necessary, setup already asks for you to `Enter your talawa-api endpoint` using the `REACT_APP_TALAWA_URL`  value.

File: src/setup/askForRemoteUrl/askForRemoteUrl.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import inquirer from 'inquirer';
+
+export const validateUrl = (input: string): boolean => {
+  const ans =
+    input.includes(':') && input.indexOf(':') == input.lastIndexOf(':');
+  const urlRegex = /^(https?|ftp):\/\/[^\s/$.?#].[^\s]*$/;
+  return ans && urlRegex.test(input);
```
Comment (palisadoes) [inline src/setup/askForRemoteUrl/askForRemoteUrl.ts:7]: This not necessary, setup already asks for you to `Enter your talawa-api endpoint` using the `REACT_APP_TALAWA_URL`  value.

**Review by palisadoes — CHANGES_REQUESTED** [review]

File: src/components/Advertisements/core/AdvertisementEntry/AdvertisementEntry.test.tsx (hunk 2 above)
Comment (aarishshahmohsin) [reply src/components/Advertisements/core/AdvertisementEntry/AdvertisementEntry.test.tsx:318]: this test was redundant and was causing problems while testing. Removing it did not have an impact on the branch and line coverage.

File: setup.ts (hunk 1 above)
Comment (aarishshahmohsin) [reply setup.ts:66]: This is not asking for API endpoint, but for solving the problem.
If we are accessing the admin portal from the network and the API endpoint contains "localhost", it causes the error. So asking for the remote URL is to change the URL into a URL that can be accessed without any problem.

**Comment by aarishshahmohsin** [comment]
@palisadoes this is to fix the immediate error, the best practice would have been to give can error if the admin portal is being accessed remotely and the API endpoint contains "localhost"

**Comment by palisadoes** [comment]
So you are saying we are fixing to problems:

1. Admin not running on localhost
2. API not running on localhost

This requires the different prompts in setup. This was also why you asked the question about automatically determining the value because you were referring to the non localhost Admin scenario. I didn't understand the context of the question.

We need to solve for both scenarios.

It would be greatly preferred to make the Admin app automatically get its Admin related information. For example some browsers could be addressing Admin on IPv4 and others on IPv6 which would make a configured value for this inappropriate. Some could be addressing the app via network address translation (NAT) which would add more unpredictability.

If that's not possible, which I doubt, then the prompt needs to be clearer. It was not apparent that this was related to the Admin portal.

**Comment by aarishshahmohsin** [comment]
> So you are saying we are fixing to problems:
> 
>     1. Admin not running on localhost
> 
>     2. API not running on localhost
> 
> 
> This requires the different prompts in setup. This was also why you asked the question about automatically determining the value because you were referring to the non localhost Admin scenario. I didn't understand the context of the question.
> 
> We need to solve for both scenarios.
> 
> It would be greatly preferred to make the Admin app automatically get its Admin related information. For example some browsers could be addressing Admin on IPv4 and others on IPv6 which would make a configured value for this inappropriate. Some could be addressing the app via network address translation (NAT) which would add more unpredictability.
> 
> If that's not possible, which I doubt, then the prompt needs to be clearer. It was not apparent that this was related to the Admin portal.

@palisadoes Rethinking, I think the best approach would be to set the REACT_APP_TALAWA_URL to the remote network link instead of this workaround on the admin side. So, if the developer wants to access the portal remotely, they will set the API link for example "http://192.168.1.4:4000/graphql" instead of "http://localhost:4000/graphql", which would fix the error without needing this PR.
This could be included in the documentation and the setup prompt asking the API URL and then this PR's workaround can be scrapped. It would be the cleanest solution.

**Comment by palisadoes** [comment]
Not necessarily, this PR may still be valid.

Please test this out. The issue was originally created because things broke when:

1. Admin and API were on different hosts or
2. Admin was accessed from a remote system, not localhost

**Comment by aarishshahmohsin** [comment]
> Not necessarily, this PR may still be valid.
> 
> Please test this out. The issue was originally created because things broke when:
> 
>     1. Admin and API were on different hosts or
 This PR did not impact this case
>     2. Admin was accessed from a remote system, not localhost
This PR is fixing this only case.

**Comment by palisadoes** [comment]
Does this PR still need to be modified to a documentation only update?

**Comment by aarishshahmohsin** [comment]
> Does this PR still need to be modified to a documentation only update?

It would avoid having unnecessary additions to the code.

**Comment by palisadoes** [comment]
Please make the documentation modifications the core of this PR

**Review by palisadoes — APPROVED** [review]

**Pull request merged**